src/base/linalg/owl_base_linalg_generic.ml

```diff
@@ -191,4 +194,192 @@ let linsolve_gauss a b =
+      for j = k + 1 to n - 1 do
+        let prev = M.get lu [| i; j |] in
+        M.set lu [| i; j |] (prev -. (!temp *. M.get lu [| k; j |]))
+      done
+    done
+  done;
+  lu, indx, !d
+
+
+(* LU decomposition, return L, U, and permutation vector *)
+let lu a =
+  let k = M.kind a in
+  let lu, indx, _ = _lu_base a in
+  let n = (M.shape lu).(0) in
+  let m = (M.shape lu).(1) in
+  assert (n = m && n >= 2);
+  let l = M.eye k n in
+  for r = 1 to n - 1 do
+    for c = 0 to r - 1 do
+      let v = M.get lu [| r; c |] in
+      M.set l [| r; c |] v;
+      M.set lu [| r; c |] 0.
+    done
+  done;
+  l, lu, indx
+
+
+let _lu_solve_vec a b =
+  assert (Array.length (M.shape b) = 1);
+  let n = (M.shape a).(0) in
+  if (M.shape b).(0) <> n then failwith "LUdcmp::solve bad sizes";
+  let ii = ref 0 in
+  let sum = ref 0. in
+  let x = M.copy b in
+  let lu, indx, _ = _lu_base a in
+  for i = 0 to n - 1 do
+    let ip = indx.(i) in
+    sum := M.get x [| ip |];
+    M.set x [| ip |] (M.get x [| i |]);
+    if !ii <> 0
+    then
+      for j = !ii - 1 to i - 1 do
+        sum := !sum -. (M.get lu [| i; j |] *. M.get x [| j |])
+      done
+    else if !sum <> 0.
+    then ii := !ii + 1;
+    M.set x [| i |] !sum
+  done;
+  for i = n - 1 downto 0 do
+    sum := M.get x [| i |];
+    for j = i + 1 to n - 1 do
+      sum := !sum -. (M.get lu [| i; j |] *. M.get x [| j |])
+    done;
+    M.set x [| i |] (!sum /. M.get lu [| i; i |])
+  done;
+  x
+
+
+(* Linear equation solution by LU decomposition.
+ * Input matrix: a[n][n], b[n][m];
+ * Output: ``x``, so that ax = b. *)
+let linsolve_lu a b =
+  let dims_a, dims_b = M.shape a, M.shape b in
+  let _, _ = _check_is_matrix dims_a, _check_is_matrix dims_b in
+  assert (dims_a.(0) = dims_a.(1));
+  let m = dims_b.(1) in
+  let b = M.copy b in
+  for j = 0 to m - 1 do
+    let vec = M.get_slice [ []; [ j ] ] b |> M.flatten in
+    let x = _lu_solve_vec a vec in
+    M.set_slice [ []; [ j ] ] b x
+  done;
+  b
+
+
+(* Matrix inverse *)
+let inv a =
+  let dims_a = M.shape a in
+  _check_is_matrix dims_a |> ignore;
+  assert (dims_a.(0) = dims_a.(1));
+  let n = dims_a.(0) in
+  let b = M.eye (M.kind a) n in
+  linsolve_lu a b
+
+
+(* Determinant of matrix a *)
+let det a =
+  let dims_a = M.shape a in
+  _check_is_matrix dims_a |> ignore;
+  assert (dims_a.(0) = dims_a.(1));
+  let n = dims_a.(0) in
+  let lu, _, sign = _lu_base a in
+  let big = ref sign in
+  for i = 0 to n - 1 do
+    big := !big *. M.get lu [| i; i |]
+  done;
+  !big
+
+
+(* Solver for tridiagonal matrix
+ * Input: a[n], b[n], c[n], which together consit the tridiagonal matrix A, and the right side vector r[n]. Return: x[n].
+ *)
+
+let tridiag_solve_vec a b c r =
+  let n = Array.length a in
+  let n1 = Array.length b in
+  let n2 = Array.length c in
+  assert (n = n1 && n = n2);
+  if b.(0) = 0.
+  then (
+    Owl_log.error
+      "Error in tridiag_solve_vec: zero element at the beginning of diagonal vector.";
+    exit 1);
```

Comment:
As a rule of thumb no library should ever call exit, raising an error here (and below) should be preferred imho. Imagine running this in a julyter notebook, you could try to recover and do something else with an error. With exit you need to restart from the beginning

Reply:
Ouch! A good catch! @jzstark can you fix it?

Reply:
Thanks for pointing it out! I didn't think about this issue. I'll fix it with a new PR soon.